sdk/go/pulumi/context.go

```diff
@@ -556,8 +735,12 @@ func (ctx *Context) registerResource(
 	// Merge providers.
 	providers := mergeProviders(t, options.Parent, options.Provider, options.Providers)
 
+	// Get the provider for the resource.
+	provider := getProvider(t, options.Provider, providers)
```

Comment:
This seems like a bugfix for provider not being pulled out of providers map when options.Provider is nil?

Reply:
It's so I can save the `provider` in the ResourceState, so I can use it subsequently inside `Call` when a `self` argument is passed to it, to pass along any explicit provider or version for the resource. That way, we don't have to expose `InvokeOptions` on a method to be able to specify those things, instead we can just implicitly use the provider/version of the resource.